File: 3팀/최유정/week20_graph/assignment/BJ2468.py
from collections import deque

n = int(input())
area = []
max_height = 0
for _ in range(n):
    a = list(map(int, input().split(" ")))
    max_height = max(max(a), max_height)
    area.append(a)

dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]

def check_area(visited, n , rain, area, j, k):
    q = deque([])
    q.append([j, k])
    while(q):
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < n:
                if area[nx][ny] > rain and not visited[nx][ny]:
                    visited[nx][ny] = True
                    q.append((nx, ny))

result = 0
for i in range(max_height):
    visited = [[False]*n for _ in range(n)]
    cnt = 0
    for j in range(n):
        for k in range(n):
            if area[j][k] > i and not visited[j][k]:
                check_area(visited, n, i, area, j, k)
                cnt += 1
    result = max(result, cnt)

print(result)

File: 3팀/이가은/week20/assignment/BJ2468.java
package week20.assignment;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

/**
 * 16792KB	264ms
 */
public class BJ2468 {
    static int N;
    static int[][] grid;
    static int[][] visited;
    static int maxHeight = Integer.MIN_VALUE;

    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};
    static boolean in_range(int x, int y){
        return x >= 0 && x < N && y >= 0 && y < N;
    }
    static void dfs(int x, int y){
        for (int i=0; i<4; i++){
            int new_x = x + dx[i];
            int new_y = y + dy[i];

            if (in_range(new_x, new_y) && visited[new_x][new_y] == 0){
                visited[new_x][new_y] = 1;
                dfs(new_x, new_y);
            }
        }
    }
    static int run(){
        int cnt = 0;
        for (int i=0; i<N; i++){
            for (int j=0; j<N; j++){
                if (visited[i][j] == 0){
                    dfs(i, j);
                    cnt += 1;
                }
            }
        }
        return cnt;
    }
    // 침수된 지역 체크하는 함수
    static void setGrid(int height){
        for (int i=0; i<N; i++){
            for (int j = 0; j < N; j++) {
                if (grid[i][j] <= height) {
                    visited[i][j] = 2;
                }
            }
        }
    }

    static void initialize(){
        for (int i=0; i<N; i++){
            for (int j = 0; j < N; j++) {
                visited[i][j] = 0;
            }
        }
    }
    public static void main(String[] args) throws IOException {
        // 입력 받기
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        grid = new int[N][N];
        visited = new int[N][N];
        for (int i=0; i<N; i++){
            String[] input = br.readLine().split(" ");
            for (int j=0; j<N; j++) {
                grid[i][j] = Integer.parseInt(input[j]);
                maxHeight = Math.max(maxHeight, grid[i][j]);
            }
        }

        int temp;
        int maxSafe = 0;
        for (int i=0; i<maxHeight; i++){
            initialize();
            setGrid(i);
            temp = run();
            maxSafe = Math.max(maxSafe, temp);
        }

        System.out.println(maxSafe);
    }
}
